src/actor/Claimer.ts

```diff
@@ -0,0 +1,186 @@
+import Identity from '../identity/Identity'
+import IRequestForAttestation from '../types/RequestForAttestation'
+import { factory as LoggerFactory } from '../config/ConfigLog'
+import Credential from '../credential/Credential'
+import PublicAttesterIdentity from '../attesteridentity/PublicAttesterIdentity'
+
+const log = LoggerFactory.getLogger('Claimer')
+
+function noNulls<T>(array: Array<T | null>): array is T[] {
+  return array.every(c => c !== null)
+}
+
+function parseArgsForKilt(args: string[]): string[] {
+  return args
+    .map(arg => arg.replace('claim.contents.', ''))
+    .filter((arg: string) => {
+      if (arg.match(/claim\.cTypeHash/)) {
+        log.warn('Cannot remove cTypeHash from claim.')
+        return false
+      }
+      return true
+    })
+}
+
+/**
+ * [ASYNC] Creates a presentation for an arbitrary amount of [[Credential]]s which can be verified in [[verifyPresentation]].
+ *
+ * @param identity The Claimer [[Identity]] which owns the [[Credential]]s.
+ * @param request The message which represents multiple [[CType]]s, [[IRequestClaimsForCTypes]]s and whether privacy
+ * enhancement is supported.
+ * @param credentials The [[Credential]]s which should be verified.
+ * @param attesterPubKeys The privacy enhanced public keys of all [[AttesterIdentity]]s which signed the [[Credential]]s.
+ * @param forcePE A boolean to force privacy enhancement.
+ * @returns A message which represents either an array of [[AttestedClaim]]s if privacy enhancement is not supported
+ * or a CombinedPresentation. Both of these options can be verified.
+ */
+export async function createPresentation(
+  identity: Identity,
+  request: IRequestClaimsForCTypes,
+  credentials: Credential[],
+  attesterPubKeys: PublicAttesterIdentity[],
+  forcePE = true
+): Promise<ISubmitClaimsForCTypes> {
+  if (!request.content.allowPE && forcePE) {
+    throw new Error(
+      'Verifier requested public presentation, but privacy enhancement was forced.'
+    )
+  }
+
+  if (request.content.allowPE) {
+    const peCreds = credentials.map(c => c.privacyCredential)
+    if (!noNulls(peCreds)) {
+      throw new Error('Missing privacy enhanced credential')
+    }
+    const gabiPresentation = await identity.claimer.buildCombinedPresentation({
+      credentials: peCreds,
+      combinedPresentationReq: request.content.peRequest,
+      attesterPubKeys: attesterPubKeys.map(
+        (ai: PublicAttesterIdentity) => ai.publicGabiKey
+      ),
+    })
+    return {
+      type: MessageBodyType.SUBMIT_CLAIMS_FOR_CTYPES_PE,
+      content: gabiPresentation,
+    }
+  }
+
+  const requestedAttributes = request.content.peRequest
+    .getRequestedProperties()
+    .map((propsPerClaim: any[]) => parseArgsForKilt(propsPerClaim))
+
+  return {
+    type: MessageBodyType.SUBMIT_CLAIMS_FOR_CTYPES_PUBLIC,
+    content: credentials.map((cred, i) => {
+      const allAttrs = cred.getAttributes()
+      requestedAttributes[i].forEach((attr: any) => allAttrs.delete(attr))
```

Comment:
I disagree we should move to `cred.createPresentation` as this would add features to it, that we do not want in a public presentation. Therefore, I propose to keep this in `Claimer.createPresentation`. I added a whitelist function in 138c471.

Reply:
@wischli the credential class is private for the claimer. The feature would not be added to the public presentation, but to the private credential?

Reply:
I withdraw what I said as I misunderstood something. I moved this to `cred.createPresentation` as requested in 08da080.

Reply:
that looks much nicer now